File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
    struct Node{
        int val;
        int count  = 0;
        Node* next = nullptr;

        Node(){++count;};
        Node(int val_){
            val = val_;
            ++count;
        };
    };

public:
    stack<int> st;
    Node* min = nullptr;

    MinStack() {
        
    }
    
    void push(int val) {
        st.push(val);
        if (!min){
            min = new Node(val);
        } else if (val == min->val){
            ++min->count;
        } else if (val < min->val){
            Node* tmp = min;
            min = new Node(val);
            min->next = tmp;
        }
    }
    
    void pop() {
        if (min->val == st.top() && --min->count == 0)
            min = min->next;
        st.pop();
    }
    
    int top() {
        return st.top();
    }
    
    int getMin() {
        return min->val;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 0155-min-stack/0155-min-stack.py
class MinStack:

    def __init__(self):
        self.st = deque()
        self.mins = deque()

    def push(self, value: int) -> None:
        self.st.append(value)
        if len(self.mins) == 0 or self.mins[-1] >= value:
            self.mins.append(value)

    def pop(self) -> None:
        val = self.st[-1]
        self.st.pop()
        if self.mins[-1] == val:
            self.mins.pop()

    def top(self) -> int:
        return self.st[-1]

    def getMin(self) -> int:
        return self.mins[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(value)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()